proposals/dao/fip_99.ts

```diff
@@ -0,0 +1,137 @@
+// ensuring contracts have a specific state, etc.
+const setup: SetupUpgradeFunc = async (addresses, oldContracts, contracts, logging) => {
+  console.log(`No actions to complete in setup for fip${fipNumber}`);
+};
+
+// Tears down any changes made in setup() that need to be
+// cleaned up before doing any validation checks.
+const teardown: TeardownUpgradeFunc = async (addresses, oldContracts, contracts, logging) => {
+  console.log(`No actions to complete in teardown for fip${fipNumber}`);
+};
+
+// Run any validations required on the fip using mocha or console logging
+// IE check balances, check state of contracts, etc.
+const validate: ValidateUpgradeFunc = async (addresses, oldContracts, contracts, logging) => {
+  // Ensure that the global rate limited minter has the MINTER role & the noncustodialpriceboundpsm has the PCV_CONTROLLER role
+  await contracts.core.hasRole(ethers.utils.id('MINTER_ROLE'), addresses.raiPriceBoundPSM);
+  await contracts.core.hasRole(ethers.utils.id('PCV_CONTROLLER_ROLE'), addresses.raiPCVDripController);
+
+  // Ensure that the rai aave pcv deposit has >= 10m RAI
+  expect((await contracts.aaveRaiPCVDeposit.balance()).gte(ethers.constants.WeiPerEther.mul(10_000_000)));
+
+  // Ensure that minting is paused on the rai psm
+  expect(await contracts.raiPriceBoundPSM.mintPaused()).to.equal(true);
+
+  // Ensure that redeem is not paused on the rai psm
+  expect(await contracts.raiPriceBoundPSM.redeemPaused()).to.equal(false);
+
+  // Ensure that the fuse pool 9 rai pcv deposit and the aave rai pcv deposit are whitelisted on the pcv guardian
+  expect(await contracts.pcvGuardian.isSafeAddress(addresses.rariPool9RaiPCVDeposit)).to.equal(true);
+  expect(await contracts.pcvGuardian.isSafeAddress(addresses.aaveRaiPCVDeposit)).to.equal(true);
+
+  // Call drip on the rai pcv drip controller and make sure the rai price bound psm got the drip
+  await contracts.raiPCVDripController.drip();
+
+  expect(await contracts.rai.balanceOf(contracts.raiPriceBoundPSM.address)).to.be.gt(
+    ethers.constants.WeiPerEther.mul(1_000_000)
+  );
+
+  // Query the rai oracle
+  const raiPrice = await contracts.chainlinkRaiUsdCompositeOracle.read();
+
+  // Attempt to swap some fei for rai via the rai price bound psm
+  await forceEth((await ethers.getSigners())[0].address);
+  await forceEth(contracts.optimisticTimelock.address);
+  await contracts.fei
+    .connect(await getImpersonatedSigner(contracts.optimisticTimelock.address))
+    .transfer((await ethers.getSigners())[0].address, ethers.constants.WeiPerEther.mul(2_000_000));
+  await contracts.fei.approve(contracts.raiPriceBoundPSM.address, ethers.constants.MaxUint256);
+  await contracts.raiPriceBoundPSM.redeem(
+    addresses.feiDAOTimelock,
+    ethers.constants.WeiPerEther.mul(1_000_000),
+    ethers.constants.WeiPerEther.mul(200_000)
+  );
+  expect(await contracts.rai.balanceOf(addresses.feiDAOTimelock)).to.be.gt(ethers.constants.WeiPerEther.mul(200_000));
```

Comment:
All of this should go in an e2e test, not the validate function.

Reply:
I think it's okay to have some overlap

Reply:
I prefer separating concerns to make things more clean